osgp-adapter-domain-smartmetering/src/main/java/com/alliander/osgp/adapter/domain/smartmetering/application/services/MBusGatewayService.java

```
@@ -0,0 +1,213 @@
+public class MBusGatewayService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(MBusGatewayService.class);
+    private static final String FUNCT_EXP_MSG_TEMPLATE = "channel: %d, deviceId: %d, manufacturerId: %d, version: %d, type: %d, \n";
+
+    @Autowired
+    @Qualifier(value = "domainSmartMeteringOutgoingOsgpCoreRequestMessageSender")
+    private OsgpCoreRequestMessageSender osgpCoreRequestMessageSender;
+
+    @Autowired
+    private SmartMeterRepository smartMeteringDeviceRepository;
+
+    @Autowired
+    private WebServiceResponseMessageSender webServiceResponseMessageSender;
+
+    @Autowired
+    private DomainHelperService domainHelperService;
+
+    public MBusGatewayService() {
+        // Parameterless constructor required for transactions...
+    }
+
+    /**
+     * @param deviceMessageMetadata
+     *            the metadata of the message, including the correlationUid, the
+     *            deviceIdentification and the organisation
+     * @param requestData
+     *            the requestData of the message, including the identificatin of
+     *            the m-bus device and the channel
+     */
+    public void coupleMbusDevice(final DeviceMessageMetadata deviceMessageMetadata,
+            final CoupleMbusDeviceRequestData requestData) throws FunctionalException {
+
+        final String deviceIdentification = deviceMessageMetadata.getDeviceIdentification();
+        final String mbusDeviceIdentification = requestData.getMbusDeviceIdentification();
+
+        LOGGER.debug("coupleMbusDevice for organisationIdentification: {} for gateway: {}, m-bus device {} ",
+                deviceMessageMetadata.getOrganisationIdentification(), deviceIdentification, mbusDeviceIdentification);
+
+        try {
+            final SmartMeter gatewayDevice = this.domainHelperService.findSmartMeter(deviceIdentification);
+            final SmartMeter gasMeterDevice = this.domainHelperService.findSmartMeter(mbusDeviceIdentification);
+
```

When using the principle of 'Fail Early'; this is the moment we can do this check.  I think the check should stay in this place as it was before.